state-chain/chains/src/sol/instruction_builder.rs

```diff
@@ -0,0 +1,313 @@
+		sol_tx_core::{
+			bpf_loader_instructions::set_upgrade_authority,
+			compute_budget::ComputeBudgetInstruction,
+			program_instructions::{ProgramInstruction, SystemProgramInstruction, VaultProgram},
+		},
+		SolAccountMeta, SolAddress, SolAmount, SolAsset, SolComputeLimit, SolInstruction,
+		SolPubkey,
+	},
+	DepositChannel, TransferAssetParams,
+};
+
+pub struct SolanaInstructionBuilder;
+
+/// TODO: Implement Compute Limit calculation. pro-1357
+const COMPUTE_LIMIT: SolComputeLimit = 300_000u32;
+
+impl SolanaInstructionBuilder {
+	/// Finalize a Instruction Set. This should be internally called after a instruction set is
+	/// complete. This will add some extra instruction required for the integrity of the Solana
+	/// Transaction.
+	///
+	/// Returns the finished Instruction Set to construct the SolTransaction.
+	fn finalize(
+		mut instructions: Vec<SolInstruction>,
+		nonce_account: SolPubkey,
+		agg_key: SolPubkey,
+		compute_price: SolAmount,
+	) -> Vec<SolInstruction> {
+		// TODO: implement compute limit calculation
+		let compute_limit = COMPUTE_LIMIT;
+
+		let mut final_instructions = vec![
+			SystemProgramInstruction::advance_nonce_account(&nonce_account, &agg_key),
+			ComputeBudgetInstruction::set_compute_unit_price(compute_price),
+			ComputeBudgetInstruction::set_compute_unit_limit(compute_limit),
+		];
+
+		final_instructions.append(&mut instructions);
+
+		final_instructions
+	}
+
+	/// Create an instruction set to fetch from each `deposit_channel` being passed in.
+	/// Used to batch fetch from multiple deposit channels in a single transaction.
+	pub fn fetch_from(
+		deposit_channels: Vec<DepositChannel<Solana>>,
+		vault_program_data_account: SolAddress,
+		system_program_id: SolAddress,
+		agg_key: SolAddress,
+		nonce_account: SolAddress,
+		compute_price: SolAmount,
+	) -> Vec<SolInstruction> {
+		let instructions = deposit_channels
+			.into_iter()
+			.map(|deposit_channel| match deposit_channel.asset {
+				SolAsset::Sol => ProgramInstruction::get_instruction(
+					&VaultProgram::FetchNative {
+						seed: deposit_channel.channel_id.to_le_bytes().to_vec(),
+						bump: deposit_channel.state,
+					},
+					vec![
+						SolAccountMeta::new_readonly(vault_program_data_account.into(), false),
+						SolAccountMeta::new(agg_key.into(), true),
+						SolAccountMeta::new(deposit_channel.address.into(), false),
+						SolAccountMeta::new_readonly(system_program_id.into(), false),
+					],
+				),
+			})
+			.collect::<Vec<_>>();
+
+		Self::finalize(instructions, nonce_account.into(), agg_key.into(), compute_price)
+	}
+
+	/// Create an instruction set to `transfer` from our Vault account to a target account.
+	pub fn transfer(
+		to: TransferAssetParams<Solana>,
+		agg_key: SolAddress,
+		nonce_account: SolAddress,
+		compute_price: SolAmount,
+	) -> Vec<SolInstruction> {
+		let instructions = match to.asset {
+			SolAsset::Sol =>
+				vec![SystemProgramInstruction::transfer(&agg_key.into(), &to.to.into(), to.amount)],
```

Comment:
Is the aggkey account not touched by the transfer call? ( @albert-llimos ?)

I would have expected a `SolAccountMeta::new(agg_key.into(), true),` here, a bit like in the other calls.

Reply:
If I recall properly the SystemProgramInstruction::transfer adds that.

Reply:
Ok yeah it's a system call, not a contract call.